[FLINK-20439][runtime] Rework `scheduleOrUpdateConsumers` to `notifyPartitionDataAvailable`

zhuzhurk opened the pull request:
## What is the purpose of the change

`Execution#scheduleOrUpdateConsumers()` is no longer needed after we have introduced pipelined region scheduling and removed legacy scheduling strategies. 

## Verifying this change

This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

tillrohrmann reviewed (approved):
Thanks for creating this PR @zhuzhurk. LGTM. +1 for merging after AZP gives green light.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -1319,28 +1319,15 @@ ResultPartitionID createResultPartitionId(final IntermediateResultPartitionID re
-	 */
-	public void scheduleOrUpdateConsumers(ResultPartitionID partitionId) throws ExecutionGraphException {
-
+	public void notifyPartitionDataAvailable(ResultPartitionID partitionId) {
```
docs are added.

zhuzhurk commented:
Thanks for the reviewing @tillrohrmann 
Sorry I just realized I made a mistake because updating an unknown input channel is still needed if BLOCKING edges exist in a pipelined region. 

e.g.
A --pipelined-->B, A --pipelined-->C, B --pipelined-->D, C --blocking-->D.

In this case, even though D knows the location of C when deployed, the corresponding input channel will be unknown because C has not finished producing the BLOCKING result partition. This is needed because reading from an uncompleted BLOCKING result partition is not supported yet.

So I reverted the changes which removed `scheduleOrUpdateConsumers()` in Execution and renamed it to `updatePartitionConsumers` to be aligned with its current responsibility. I also added a test `UpdatePartitionConsumersTest` to cover this case.

This also means that FLINK-20721 is invalid so I will close it.

tillrohrmann commented:
I see. Ok then let's keep this mechanism. Ideally this is something the result partition could control itself. Then we wouldn't have to control when to start reading the results. On the other hand, it does not hurt at the moment to have this functionality.

tillrohrmann reviewed (approved):
LGTM. +1 for merging.

zhuzhurk commented:
@flinkbot run azure
